src/base/memory_monitor_sr.cpp

```diff
@@ -69,28 +102,71 @@ namespace cubmem
     return allocated_size;
   }
 
-  std::string memory_monitor::make_tag_name (const char *file, const int line)
+  void memory_monitor::add_stat (char *ptr, const size_t size, const char *file, const int line)
+  {
+    std::string tag_name;
+    MMON_METAINFO metainfo;
+
+    // size should not be 0 because of MMON_ALLOC_META_SIZE
+    assert (size > 0);
+
+    metainfo.allocated_size = (uint64_t) size;
+    m_total_mem_usage += metainfo.allocated_size;
+
+    tag_name = make_tag_name (file, line);
+
+    std::unique_lock <std::mutex> tag_map_lock (m_tag_map_mutex);
+    auto search = m_tag_map.find (tag_name);
+    if (search != m_tag_map.end ())
+      {
+	metainfo.tag_id = search->second;
+	m_stat_map[metainfo.tag_id] += metainfo.allocated_size;
+      }
+    else
+      {
+	metainfo.tag_id = m_tag_map.size ();
+	// tag_id starts with 0
+	m_tag_map.emplace (tag_name, metainfo.tag_id);
+	m_stat_map.emplace (metainfo.tag_id, metainfo.allocated_size);
+      }
+    tag_map_lock.unlock ();
+
+    // put meta info into the allocated chunk
+    char *meta_ptr = get_meta_ptr (ptr, metainfo.allocated_size);
+    metainfo.magic_number = m_magic_number;
+    memcpy (meta_ptr, &metainfo, MMON_ALLOC_META_SIZE);
+    m_meta_alloc_count++;
+  }
+
+  void memory_monitor::sub_stat (char *ptr)
   {
-    std::string filecopy (file);
 #if defined(WINDOWS)
-    std::string target ("\\src\\");
+    size_t allocated_size = 0;
+    assert (false);
 #else
-    std::string target ("/src/");
-#endif // WINDOWS
+    size_t allocated_size = malloc_usable_size ((void *)ptr);
+#endif // !WINDOWS
 
-    // Find the last occurrence of "src" in the path
-    size_t pos = filecopy.rfind (target);
-#if !defined (NDEBUG)
-    size_t lpos = filecopy.find (target);
-    assert (pos == lpos);
-#endif // NDEBUG
+    assert (ptr != NULL);
 
-    if (pos != std::string::npos)
+    if (allocated_size >= MMON_ALLOC_META_SIZE)
       {
-	filecopy = filecopy.substr (pos + target.length ());
-      }
+	char *meta_ptr = get_meta_ptr (ptr, allocated_size);
+	const MMON_METAINFO *metainfo = (const MMON_METAINFO *) meta_ptr;
 
-    return filecopy + ':' + std::to_string (line);
+	if (metainfo->magic_number == m_magic_number)
+	  {
+	    assert ((metainfo->tag_id >= 0 && metainfo->tag_id <= m_stat_map.size()));
+	    assert (m_stat_map[metainfo->tag_id] >= metainfo->allocated_size);
+
+	    m_total_mem_usage -= metainfo->allocated_size;
+	    m_stat_map[metainfo->tag_id] -= metainfo->allocated_size;
+
+	    memset (meta_ptr, 0, MMON_ALLOC_META_SIZE);
+	    m_meta_alloc_count--;
+	    assert (m_meta_alloc_count >= 0);
```

Comment:
```suggestion
	    assert (m_total_mem_usage >= 0);
	    assert (m_meta_alloc_count >= 0);
```

Reply:
I think `assert(m_stat_map[metainfo->tag_id] >= metainfo->allocated_size)` can guarantee `m_total_mem_usage >= 0` after sub_stat but we can check another abnormal case if we add this assertion. Done.